### PR TITLE
Restructure common utils

### PR DESCRIPTION
**Pull request by kirilklein**
Add functions from common.utils to modules inside functional.
Data class is still used but will be removed shortly, so we leave it for now in common. utils.
Also, functions for conversion from epochs to num_steps were refactored.

### CODE REVIEW

File: corebehrt/common/utils.py [hunk 1]
```diff
@@ -74,23 +21,29 @@ class Data:
...
     vocabulary: Optional[Dict] = field(default=None)
     mode: Optional[str] = field(default=None)
-    
+
     def __len__(self):
         return len(self.pids)
-    
-    def copy(self) -> 'Data':
+
+    def copy(self) -> "Data":
```
Comment (mikkelfo) [inline corebehrt/common/utils.py:28]: I think we should be able to just `deepcopy(self)` and all attributes should automatically be deepcopied as well

File: corebehrt/common/utils.py [hunk 2]
```diff
@@ -100,80 +53,109 @@ def load_tensor(filename, required=False):
                 else:
                     return None
             return torch.load(filepath)
-        features = load_tensor(f'{mode}_features.pt', required=True)
-        pids = load_tensor(f'{mode}_pids.pt', required=True)
-        outcomes = load_tensor(f'{mode}_outcomes.pt')
-        censor_outcomes = load_tensor(f'{mode}_censor_outcomes.pt')
-        vocabulary = load_tensor('vocabulary.pt')
+
+        features = load_tensor(f"{mode}_features.pt", required=True)
```
Comment (mikkelfo) [inline corebehrt/common/utils.py:57]: I think we should check if filepath exists and load if exists (using `path().exists()`) rather than "loading" a file that isn't present for clarity

File: corebehrt/common/utils.py [hunk 3]
```diff
@@ -100,80 +53,109 @@ def load_tensor(filename, required=False):
...
-        return Data(features={key: [values[i] for i in indices] for key, values in self.features.items()}, 
-                        pids=[self.pids[i] for i in indices],
-                        outcomes=[self.outcomes[i] for i in indices] if self.outcomes is not None else None,
-                        censor_outcomes=[self.censor_outcomes[i] for i in indices] if self.censor_outcomes is not None else None,
-                        vocabulary=self.vocabulary,
-                        mode=mode)
-    
-    def select_data_subset_by_pids(self, pids: list, mode:str='')->'Data':
+
+    def select_data_subset_by_indices(self, indices: list, mode: str = "") -> "Data":
```
Comment (mikkelfo) [inline corebehrt/common/utils.py:87]: This should maybe be a part of the `Data` class as a `subset` method, as it only functions on this

File: corebehrt/functional/data_check.py [hunk 4]
```diff
@@ -19,3 +21,17 @@ def log_features_in_sequence(data: dict) -> None:
     logger.info("Example features: ")
     for k, v in data.features.items():
         logger.info(f"{k}: {v[0]}")
+
+
+def check_patient_counts(
```
Comment (mikkelfo) [inline corebehrt/functional/data_check.py:26]: Should this only issue a warning or is this an error?

File: corebehrt/functional/files_and_folders.py [hunk 5]
```diff
@@ -0,0 +1,14 @@
+import os
+
+
+def split_path(path_str: str) -> list:
```
Comment (mikkelfo) [inline corebehrt/functional/files_and_folders.py:4]: This can be simplified using `pathlib Path` by doing `Path(path_str).parts` to split up and extract directories (by doing `parts[:-1]`

File: corebehrt/main/pretrain.py [hunk 6]
```diff
@@ -38,8 +38,10 @@ def main_train(config_path):
     train_dataset, val_dataset = DatasetPreparer(cfg).prepare_mlm_dataset()
 
     if "scheduler" in cfg:
```
Comment (mikkelfo) [inline corebehrt/main/pretrain.py:40]: This is duplicated for finetune, so perhaps it should be wrapped in a functionality for standardization?

File: corebehrt/main/finetune_cv.py [hunk 7]
```diff
@@ -133,8 +134,11 @@ def finetune_fold(
 ) -> None:
     """Finetune model on one fold"""
     if "scheduler" in cfg:
```
Comment (mikkelfo) [inline corebehrt/main/finetune_cv.py:136]: See comment below for `pretrain.py`

**Review by mikkelfo — CHANGES_REQUESTED** [review]
Feel free to ignore my comments for the `Data` dataclass if it is getting removed, just thought I'd add my comments to the record

File: corebehrt/common/utils.py (hunk 2 above)
Comment (kirilklein) [reply corebehrt/common/utils.py:57]: Do you mean doing a check inside the load_from_directory method instead of inside the submethod? 
Since not all the files need to be present to construct the data, we check if optional files are in the directory and return None if not. What would you suggest instead?

We might figure out a better way to do it for the new dataclass.

File: corebehrt/common/utils.py (hunk 3 above)
Comment (kirilklein) [reply corebehrt/common/utils.py:87]: We won't need the method in the future

**Comment by kirilklein** [comment]
> Feel free to ignore my comments for the `Data` dataclass if it is getting removed, just thought I'd add my comments to the record

OK. I don't think we should completely get rid of it, since a dataclass also makes sense with dataframes, especially if we have outcomes and censor_outcomes in separate dfs. This will help us keep track of the patients and allow for some checks. However, most methods won't be needed in the new setup.

File: corebehrt/common/utils.py (hunk 1 above)
Comment (kirilklein) [reply corebehrt/common/utils.py:28]: Good point! We should note that for the df version of the method.

File: corebehrt/functional/data_check.py (hunk 4 above)
Comment (kirilklein) [reply corebehrt/functional/data_check.py:26]: I think this method is no longer needed. Also, we'll get rid of the ConceptLoaderLarge.
It's probably a good idea to tackle create_outcomes next to rewrite with dask

Comment (kirilklein) [reply corebehrt/functional/data_check.py:26]: Adressed in a8129132b34860ad4e39afbd1131375127d61d3f

File: corebehrt/functional/files_and_folders.py (hunk 5 above)
Comment (kirilklein) [reply corebehrt/functional/files_and_folders.py]: Addressed in 52be7d6314f3ad08240774c9eefd34fa17927768

File: corebehrt/main/pretrain.py (hunk 6 above)
Comment (kirilklein) [reply corebehrt/main/pretrain.py:40]: I'm not sure what you mean exactly. You mean the whole thing with the if statement?

**Comment by kirilklein** [comment]
@mikkelfo I've adressed your comments, except the ones for the dataclass as we'll rewrite it and get rid of most methods.
Could you please clarify this comment;
> This is duplicated for finetune, so perhaps it should be wrapped in a functionality for standardization?

File: corebehrt/common/utils.py (hunk 2 above)
Comment (mikkelfo) [reply corebehrt/common/utils.py:57]: Yeah there might be a few options to "fix" it IMO. Just feel like it's a little odd to have a bunch of load statements that might be doing
We might do it inside the load_from_directory method or maybe we can just change the `load_tensor` to `load_if_exists` or something like that to clarify. Or have a pretrain/finetune one, so we only load the expected files or something like that

File: corebehrt/main/pretrain.py (hunk 6 above)
Comment (mikkelfo) [reply corebehrt/main/pretrain.py:40]: Yes. If we have to adjust the scheduler in the same way for both pretraining and finetune, we should standardize it, rather than having it duplicates in both scripts

Comment (mikkelfo) [reply corebehrt/main/pretrain.py:40]: But now where I look at it today, I guess it's just an if_statement with a log, so might not be needed

**Review by mikkelfo — APPROVED** [review]
I don't have direct solutions to my other comments (and they might not be valid), so it's fine to merge now imo
